<a href="https://colab.research.google.com/github/sh4dhub4b4/CSE-413-Simulation_Modeling_Lab/blob/main/Lab2_adnan_061(B1).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [92]:
import numpy as np
from sympy import Matrix
import matplotlib.pyplot as plt
import seaborn as sns


In [93]:
# Task 1
# id = 1061
d1,d2=1,6
A=[
    [d1+2,d2+1],
    [2*d1,d2+2]
]
Matrix(A)

Matrix([
[3, 7],
[2, 8]])

In [94]:
# Task 2
# matA=Matrix(A)
# shapeA=matA.shape
# detA=matA.det()
# rankA=matA.rank()
# eigenvalsA=matA.eigenvals()
# invA=matA.inv() if detA!=0 else 'None'

matA = np.array(A)
shapeA = matA.shape
detA = np.linalg.det(matA)
rankA = np.linalg.matrix_rank(matA)
eigenvalsA = np.linalg.eigvals(matA)
invA = np.linalg.inv(matA) if detA!=0 else 'None'

print("Shape:", shapeA)
print("Determinant:", detA)
print("Rank:", rankA)
print("Eigenvalues:", eigenvalsA)
print("Inverse:\n")
display(Matrix(invA))

Shape: (2, 2)
Determinant: 10.000000000000002
Rank: 2
Eigenvalues: [ 1. 10.]
Inverse:



Matrix([
[ 0.8, -0.7],
[-0.2,  0.3]])

In [99]:
# Task 3

B = [row[:] for row in A]
B[0][1]+=1 #changing the (0,1) element

print("A =")
display(Matrix(A))

print("B =")
display(Matrix(B))

A =


Matrix([
[3, 7],
[2, 8]])

B =


Matrix([
[3, 8],
[2, 8]])

In [101]:
# Task 4
matB=np.array(B)
shapeB=matB.shape
detB=np.linalg.det(matB)
rankB=np.linalg.matrix_rank(matB)
eigenvalsB=np.linalg.eigvals(matB)
invB=np.linalg.inv(matB) if detB!=0 else 'None'

print("Shape:", shapeB)
print("Determinant:", detB)
print("Rank:", rankB)
print("Eigenvalues:", eigenvalsB)
print("Inverse:\n")
display(Matrix(invB))

Shape: (2, 2)
Determinant: 8.000000000000002
Rank: 2
Eigenvalues: [ 0.78300943 10.21699057]
Inverse:



Matrix([
[  1.0,  -1.0],
[-0.25, 0.375]])

# Task 5

### 1. How did the determinant change and why?

The determinant changed from **10** to **8** after increasing one value in the matrix. This happened because the determinant depends on all the elements of the matrix.

---

### 2. Did the rank change? Explain.

No, the rank did not change. Both matrices have a rank of **2** because their determinants are not zero.

---

### 3. How did the eigenvalues respond to the value change?

The eigenvalues changed slightly.

- **Matrix A:** = **10**, **1**
- **Matrix B:** = **0.78300943**,**10.21699057**

Changing one element of the matrix caused the eigenvalues to change.

---

### 4. Is B easier or harder to invert than A? Why?

Both matrices are easy to invert because both have non-zero determinants.

However, **B** has a smaller determinant than **A**, so it is slightly closer to being singular and can be a little more sensitive to numerical errors.